bin/ruby-build

```diff
@@ -1081,6 +1081,16 @@ has_broken_mac_openssl() {
   [[ "$RUBY_CONFIGURE_OPTS" != *--with-openssl-dir=* ]]
 }
 
+has_openssl_3() {
+  local openssl_version="$(openssl version 2>/dev/null || printf '#include <openssl/opensslv.h>\nOPENSSL_VERSION_TEXT\n' | cc -xc -E - 2>/dev/null | tail -n1 | tr -d '"' || true)"
+  [[ $openssl_version = "OpenSSL 3"?* ]] &&
+  [[ "$RUBY_CONFIGURE_OPTS" != *--with-openssl-dir=* ]]
```

Comment:
This line should be the first condition so the rest of the logic short-circuits if `--with-openssl-dir` is set (that way we don't have to invoke `cc`, for example)